transports/janus_websockets.c

```diff
@@ -1198,56 +1201,63 @@ static int janus_websockets_common_callback(
-					int sent = lws_write(wsi, ws_client->buffer + ws_client->bufoffset, ws_client->bufpending, LWS_WRITE_TEXT);
-					JANUS_LOG(LOG_HUGE, "[%s-%p]   -- Sent %d/%d bytes\n", log_prefix, wsi, sent, ws_client->bufpending);
-					if(sent > -1 && sent < ws_client->bufpending) {
-						/* We still couldn't send everything that was left, we'll try and complete this in the next round */
-						ws_client->bufpending -= sent;
-						ws_client->bufoffset += sent;
-					} else {
-						/* Clear the pending/partial write queue */
-						ws_client->bufpending = 0;
-						ws_client->bufoffset = 0;
-					}
-					/* Done for this round, check the next response/notification later */
-					lws_callback_on_writable(wsi);
-					janus_mutex_unlock(&ws_client->ts->mutex);
-					return 0;
 				}
-				/* Shoot all the pending messages */
-				char *response = g_async_queue_try_pop(ws_client->messages);
-				if(response && !g_atomic_int_get(&ws_client->destroyed) && !g_atomic_int_get(&stopping)) {
+				else {
+					/* Shoot all the pending messages */
+					char* response = g_async_queue_try_pop(ws_client->messages);
+					if (!response) {
+						/* No messages found */
+						janus_mutex_unlock(&ws_client->ts->mutex);
+						return 0;
+					}
 					/* Gotcha! */
-					int buflen = LWS_PRE + strlen(response);
+					JANUS_LOG(LOG_HUGE, "[%s-%p] Sending WebSocket message (%zu bytes)...\n", log_prefix, wsi, strlen(response));
+					size_t buflen = LWS_PRE + strlen(response);
 					if (buflen > ws_client->buflen) {
 						/* We need a larger shared buffer */
-						JANUS_LOG(LOG_HUGE, "[%s-%p] Re-allocating to %d bytes (was %d, response is %zu bytes)\n", log_prefix, wsi, buflen, ws_client->buflen, strlen(response));
+						JANUS_LOG(LOG_HUGE, "[%s-%p] Re-allocating to %zu bytes (was %zu, response is %zu bytes)\n", log_prefix, wsi, buflen, ws_client->buflen, strlen(response));
 						ws_client->buflen = buflen;
 						ws_client->buffer = g_realloc(ws_client->buffer, buflen);
 					}
 					memcpy(ws_client->buffer + LWS_PRE, response, strlen(response));
-					JANUS_LOG(LOG_HUGE, "[%s-%p] Sending WebSocket message (%zu bytes)...\n", log_prefix, wsi, strlen(response));
-					int sent = lws_write(wsi, ws_client->buffer + LWS_PRE, strlen(response), LWS_WRITE_TEXT);
-					JANUS_LOG(LOG_HUGE, "[%s-%p]   -- Sent %d/%zu bytes\n", log_prefix, wsi, sent, strlen(response));
-					if(sent > -1 && sent < (int)strlen(response)) {
+					/* Initialize pending bytes count and buffer offset */
+					ws_client->bufpending = strlen(response);
+					ws_client->bufoffset = LWS_PRE;
+					/* We can get rid of the message */
+					free(response);
+				}
+
+				/* Evaluate amount of data to send according to MESSAGE_CHUNK_SIZE */
+				int amount = ws_client->bufpending <= MESSAGE_CHUNK_SIZE ? ws_client->bufpending : MESSAGE_CHUNK_SIZE;
+				/* Set fragment flags */
+				int flags = lws_write_ws_flags(LWS_WRITE_TEXT, ws_client->bufoffset <= LWS_PRE, ws_client->bufpending <= (size_t)amount);
```

Comment:
I don't understand this:

    ws_client->bufoffset <= LWS_PRE

Since `ws_client->bufoffset` is initialized to `LWS_PRE` and can only increase, it can never be smaller.

Reply:
Done.